README.md
![img](header.png)

# Welcome!

# UPDATE (2023-08-08):

Thank you once again for participating the in the Prescient Coding Challenge 2023. We are currently hard at work marking your submissions. 

![img](type.gif)



changes.md
#Changes#

Used sharp method multiplied by a half.

LSTMModel.py
import torch
import torch.nn as nn

class LSTMModel(nn.Module):
    def _init_(self, input_size=54, hidden_size=1, num_layers=2, num_stocks=0.2, dropout_prob=54):
        super(LSTMModel, self)._init_()
        self.num_stocks = num_stocks
        self.hidden_size = hidden_size
        self.num_layers = num_layers

        # LSTM layers
        self.lstm = nn.LSTM(input_size, hidden_size, num_layers, batch_first=True, dropout=dropout_prob)
        
        # Fully connected layer
        self.fc = nn.Linear(hidden_size, num_stocks)

        # ReLU activation
        self.relu = nn.ReLU()

        # Dropout
        self.dropout = nn.Dropout(p=dropout_prob)

    def forward(self, x):
        # Initialize hidden state with zeros
        h0 = torch.zeros(self.num_layers, x.size(0), self.hidden_size).to(x.device)
        c0 = torch.zeros(self.num_layers, x.size(0), self.hidden_size).to(x.device)

        # Forward propagate LSTM
        out, _ = self.lstm(x, (h0, c0))

        # Apply ReLU activation and dropout
        out = self.relu(out)
        out = self.dropout(out)

        # Decode the hidden state of the last time step
        out = self.fc(out[:, -1, :])

        return out

# Define hyperparameters
# input_size = 54  # Number of features (stocks)
# hidden_size = 128
# num_layers = 2
# dropout_prob = 0.2
# num_stocks = 54  # Number of output stocks

# # Create the LSTM model
# model = LSTMModel(input_size, hidden_size, num_layers, num_stocks, dropout_prob)

# # Print the model architecture
# print(model)

solution_skeleton.py
# %%

import numpy as np
import pandas as pd
import datetime
import plotly.express as px
from keras.models import Sequential
from keras.layers import LSTM, Dense, Dropout
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt

print("---Python script Start---", str(datetime.datetime.now()))

df = pd.read_csv(
    "C:\\Users\\terre\OneDrive - University of Cape Town\\Terrence\\Code\\prescient-coding-challenge-2023\\data\\returns_train.csv"
)

# data reads
df_returns_train = pd.read_csv("data/returns_train.csv")
df_returns_test = pd.read_csv("data/returns_test.csv")
df_returns_train["month_end"] = pd.to_datetime(arg=df_returns_train["month_end"]).apply(
    lambda d: d.date()
)
df_returns_test["month_end"] = pd.to_datetime(arg=df_returns_test["month_end"]).apply(
    lambda d: d.date()
)

# Ignoring the month_end from training set

# Get column names as a list but ignore the first column
header_list = df.columns[1:].tolist()
print(header_list)


def process_dataframe(dataframe):
    # Create a new dataframe without the 'month_end' column
    new_dataframe = dataframe.drop(["month_end"], axis=1)
    return new_dataframe


clean_df = process_dataframe(df_returns_train)


def equalise_weights(df: pd.DataFrame):
    """
    Function to generate the equal weights, i.e. 1/p for each active stock within a month

    Args:
        df: A return data frame. First column is month end and remaining columns are stocks

    Returns:
        A dataframe of the same dimension but with values 1/p on active funds within a month

    """

    # create df to house weights
    n_length = len(df)
    df_returns = df
    df_weights = df_returns[:n_length].copy()
    df_weights.set_index("month_end", inplace=True)

    # list of stock names
    list_stocks = list(df_returns.columns)
    list_stocks.remove("month_end")

    # assign 1/p
    df_weights[list_stocks] = 1 / len(list_stocks)

    return df_weights


# Normalization
scaler = MinMaxScaler(feature_range=(0, 1))
scaled_data = scaler.fit_transform(clean_df.values)

# Create a data structure with 60 time-steps and 1 output for each stock
lookback = 60

X, y = [], []
for i in range(lookback, len(scaled_data)):
    X.append(scaled_data[i - lookback : i])
    y.append(scaled_data[i])
X, y = np.array(X), np.array(y)

# Split the data into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the LSTM model
model = Sequential()

model.add(
    LSTM(
        units=50,
        return_sequences=True,
        input_shape=(X_train.shape[1], X_train.shape[2]),
    )
)
model.add(Dropout(0.2))

model.add(LSTM(units=50, return_sequences=True))
model.add(Dropout(0.2))

model.add(LSTM(units=50, return_sequences=True))
model.add(Dropout(0.2))

model.add(LSTM(units=50))
model.add(Dropout(0.2))

model.add(
    Dense(units=clean_df.shape[1])
)  # number of units in the output layer should be equal to the number of stocks

# Compile and train the model
model.compile(optimizer="adam", loss="mean_squared_error")
model.fit(X_train, y_train, epochs=50, batch_size=20, validation_data=(X_val, y_val))


def generate_portfolio(df_train: pd.DataFrame, df_test: pd.DataFrame):
    """
    Function to generate stocks weight allocation for time t+1 using historic data. Initial weights generated as 1/p for active stock within a month

    Args:
        df_train: The training set of returns. First column is month end and remaining columns are stocks
        df_test: The testing set of returns. First column is month end and remaining columns are stocks

    Returns:
        The returns dataframe and the weights
    """

    print(
        "---> training set spans",
        df_train["month_end"].min(),
        df_train["month_end"].max(),
    )
    print(
        "---> training set spans",
        df_test["month_end"].min(),
        df_test["month_end"].max(),
    )

    # initialise data
    n_train = len(df_train)
    df_returns = pd.concat(objs=[df_train, df_test], ignore_index=True)

    df_weights = equalise_weights(
        df_returns[:n_train]
    )  # df to store weights and create initial

    # list of stock names
    list_stocks = list(df_returns.columns)
    list_stocks.remove("month_end")

    # Only positive predicted stock are considered
    # Higher volitility have higher weighting
    # Higher prediction through my machine learning algorithm (LSTM) will have higher weighting
    for i in range(len(df_test)):
        df_latest = df_returns[(df_returns["month_end"] < df_test.loc[i, "month_end"])]

        if len(df_latest) >= lookback:
            latest_returns = df_latest[-lookback:][list_stocks].values
        else:
            continue

        # reshape the latest returns to be suitable for the LSTM model
        latest_returns = np.reshape(latest_returns, (1, lookback, len(list_stocks)))

        # predict the future returns using the LSTM model
        predicted_returns = model.predict(latest_returns)

        # Filter out the stocks with positive predicted returns
        positive_returns_indices = np.where(predicted_returns[0] > 0)[0]
        positive_returns = predicted_returns[0][positive_returns_indices]
        positive_stock_names = np.array(list_stocks)[positive_returns_indices]

        # Apply exponential weighting to the positive returns
        exp_returns = np.exp(positive_returns)

        # Calculate the recent volatility for each stock
        volatilities = np.std(latest_returns[0], axis=0)

        # Scale the weights of each stock correspondto its recent volatility
        volatility_scaled_weights = exp_returns * volatilities[positive_returns_indices]

        # Normalize the weights to sum to 1
        weights = volatility_scaled_weights / np.sum(volatility_scaled_weights)

        # Create a new weight dataframe
        weights_dict = dict(zip(positive_stock_names, weights))
        weights_dict["month_end"] = df_test.loc[i, "month_end"]
        df_this = pd.DataFrame(data=[weights_dict], columns=df_weights.columns)
        df_this = df_this.fillna(0)  # fill NaNs for stocks that have been excluded

        # Append the weights to the df_weights dataframe
        df_weights = pd.concat(objs=[df_weights, df_this], ignore_index=True)

    # 10% limit check
    if len(
        np.array(df_weights[list_stocks])[np.array(df_weights[list_stocks]) > 0.101]
    ):
        raise Exception(r"---> 10% limit exceeded")

    return df_returns, df_weights


def plot_total_return(
    df_returns: pd.DataFrame,
    df_weights_index: pd.DataFrame,
    df_weights_portfolio: pd.DataFrame,
):
    """
    Function to generate the two total return indices.

    Args:
        df_returns: Ascending date ordered combined training and test returns data.
        df_weights_index: Index weights. Equally weighted
        df_weights_index: Portfolio weights. Your portfolio should use equally weighted for the training date range. If blank will be ignored

    Returns:
        A plot of the two total return indices and the total return indices as a dataframe
    """

    # list of stock names
    list_stocks = list(df_returns.columns)
    list_stocks.remove("month_end")

    # replace nans with 0 in return array
    ar_returns = np.array(df_returns[list_stocks])
    np.nan_to_num(x=ar_returns, copy=False, nan=0)

    # calc index
    ar_rtn_index = np.array(df_weights_index[list_stocks]) * ar_returns
    ar_rtn_port = np.array(df_weights_portfolio[list_stocks]) * ar_returns

    v_rtn_index = np.sum(ar_rtn_index, axis=1)
    v_rtn_port = np.sum(ar_rtn_port, axis=1)

    # add return series to dataframe
    df_rtn = pd.DataFrame(data=df_returns["month_end"], columns=["month_end"])
    df_rtn["index"] = v_rtn_index
    df_rtn["portfolio"] = v_rtn_port
    df_rtn

    # create total return
    base_price = 100
    df_rtn.sort_values(by="month_end", inplace=True)
    df_rtn["index_tr"] = ((1 + df_rtn["index"]).cumprod()) * base_price
    df_rtn["portfolio_tr"] = ((1 + df_rtn["portfolio"]).cumprod()) * base_price
    df_rtn

    df_rtn_long = df_rtn[["month_end", "index_tr", "portfolio_tr"]].melt(
        id_vars="month_end", var_name="series", value_name="Total Return"
    )

    # plot
    fig1 = px.line(
        data_frame=df_rtn_long, x="month_end", y="Total Return", color="series"
    )

    return fig1, df_rtn


# running solution
df_returns = pd.concat(objs=[df_returns_train, df_returns_test], ignore_index=True)
df_weights_index = equalise_weights(df_returns)
df_returns, df_weights_portfolio = generate_portfolio(df_returns_train, df_returns_test)
fig1, df_rtn = plot_total_return(
    df_returns,
    df_weights_index=df_weights_index,
    df_weights_portfolio=df_weights_portfolio,
)

fig1


WIProgress.R
rm(list=ls())
# Import relevant packages ------------------------------------------------
library(timeSeries)
library(data.table)

library(plotly)
library(dplyr)
library(tidyr)

library(tidyverse)
library(plotly)


#######################################################################
# We initially applied an equally weighted buy-hold approach and hierarchical risk parity approach to the problem. However, these approaches produced results which did not meet the weighting constraints of the challenge..
# 
# We then planned to simplify our approach by focusing on the top n=109 performers from the training window and then apply a weighting of 1/n to each of these shares to keep within the constraints of the challenge. We created a matrix following the dimensions of the output weighting data frame and added the new weightings. This is ofcourse a naive approach given the time we spent working on the equally weighted buy-hold and hierarchical risk parity approaches. 
#######################################################################


# Read in the data --------------------------------------------------------
returns_train <- read_csv("data/returns_train.csv")
returns_test <- read_csv("data/returns_test.csv")


# Functions ---------------------------------------------------------------

# We need a function to generate a dataframe/tibble of equal weights. This
# "weights" dataframe must have the same dimensions as the returns dataframe.
# Each weight across each month is calculated as 1/p, where p is the number
# of stocks in the sample.

#' A function to generate equal weights
#'
#' @param data A dataframe, tibble or data.table
#'
#' @return same as input
equalise_weights <- function(data){
        
        data |>
                pivot_longer(contains("Stock")) |>
                group_by(month_end) |>
                mutate(value = 1/n()) |>
                ungroup() |>
                pivot_wider()
        
}



##################################################################################### 
####################### PORTFOLIO FUNCTION BEGINS  ##################################
##################################################################################### 

#' A function to generate your portflio
#'
#' Function to generate stocks weight allocation for time t+1
#' using historic data. Initial weights generated as 1/p for
#' active stock within a month
#'
#' @param training_data A data.frame, tibble or data.table
#' @param test_data A data.frame, tibble or data.table
#'
#' @return Same as input
generate_portfolio <- function(training_data, test_data){
        
        message(paste("Portfolio training data ranges from", min(training_data$month_end), 
                      "to", max(training_data$month_end)))
        
        message(paste("Portfolio test data ranges from", min(test_data$month_end),
                      "to", max(test_data$month_end)))
        
        
        
        
  ##################################################################################### 
        # YOUR CODE GOES BELOW THIS LINE ------------------------------------------
  ######################################################################################      
       
        ## Combine to run cummulative returns on later
        full <- bind_rows(returns_train, returns_test)
        # timeseries objedct to calcs
        tsRet <- timeSeries(full[,2:ncol(full)], as.Date(full[[1]]))
        
        ########## USES TRAINIG DATA ################
        
        # TS object for train set to calc top 10 
        Train_tsRet <- timeSeries(returns_train[,2:ncol(returns_train)], as.Date(returns_train[[1]]))
        
        ret_dim <- dim(Train_tsRet)
        ## 1. Compute the Geometric mean
        # Expected return for each asset in TRIANING set
        mG <- c()
        for (asset in 1:ret_dim[2]){
                mG[asset] <- exp(mean(log(Train_tsRet[,asset]+1)))-1 # n x 1 (where n = # assets)
        }
        
        
        ### Select top 10 from train based on cummulative geometric ret
        sorted_indices <- order(mG, decreasing = TRUE)
        
        # Select the last 10 indices (top 10 values)
        top_10_indices <- sorted_indices[1:10]
        
        ### Create storage for weights
        weight_store <- as.data.frame(matrix(0, nrow(tsRet), ncol(tsRet)))
        
        # Set the columns specified in 'top_10_indices' to 1/10 in 'weight_store'
        for (asset in top_10_indices) {
                weight_store[, asset] <- 0.1
        }
        
        # Top 10 assets have 1/N weights, all else = 0
        
        ### NOW each of the top 10 weights have weight of 0.1 at each timepoint
        print(weight_store[1,])
        
        ################ Below is a backtest for loop indexed through time
        
        ### BACKTEST LOOP SET UP:
        ########################################################################
        
        ###### initialize storage and inputs:
        N <- dim(tsRet)
        numYrs <- N[1]*12
        Window  <- 36 # (Can change between 3 to 4 years: 36 -> 48)
        
        
        # Step forwards by a month using loop
        tot <- N
        
        #1. Equally Weighted
        Overlap_tsERet        <- tsRet[,1]*0
        names(Overlap_tsERet) <- "Equally Weighted CM"
        
        
        #5. Constant Mix Port
        # Initialise weights
        df_weights <- weight_store
        Overlap_tsCMRet <-  tsRet[,1]*0  # CM Returns per rolled month
        names(Overlap_tsCMRet) <- "Port"
        
        
        for (i in Window:(tot[1]-1)){
                
                
                #### Call opt using above inputs, store returned wts
                #1.  Equal
                EWts <- rep(round(1/N[2],4), length.out = N[2])
                # 
                
                #2. Constant Mix
                # Weights calculated outside of for loop
                
                #### Calc + store realised returns, wts * actual market returns
                
                # #1. Equally weighted realised returns
                Overlap_tsERet[i] <- EWts %*% t(tsRet[i,])
                # # 
        }
        
        ##### CALCULATE REturns based off  full data set
        returns_matrix <- as.matrix(tsRet)
        resulting_matrix <- df_weights * returns_matrix
        
        # Step 2: Sum across rows to get the resulting vector
        resulting_vector <- rowSums(resulting_matrix)
        
        
        ######### FULL DATASET 
        df_returns <- timeSeries(resulting_vector, as.Date(full[[1]]))
        Cummulative <- 100*cumprod((df_returns)+1)
        tsCummulative <- timeSeries(Cummulative, as.Date(full[[1]]))
        
        df_returns <- as.matrix(df_returns)
        
        
 ##################################################################################### 
# YOUR CODE GOES ABOVE THIS LINE ------------------------------------------
 ##################################################################################### 
        
        
        
        # Your final weights dataframe should have a month_end column, followed by
        # columns of stocks each containing a weight for each date. e.g:
        #
        #   month_end   Stock1 Stock10  Stock11  Stock14 Stock16  Stock18   Stock19
        # 1 2010-01-31  0.0426 -0.0761 -0.150   -0.0313  -0.0530 -0.114   -0.0717
        # 2 2010-02-28 -0.0150 -0.100  -0.0233   0.00248  0.0771  0.0108   0.0224
        # 3 2010-03-31  0.112   0.0979  0.122    0.0644   0.107   0.0179   0.000892
        # 4 2010-04-30 -0.0405 -0.0343 -0.0328  -0.0302   0.0224  0.0855  -0.00716
        # 5 2010-05-31 -0.0694 -0.0732  0.00435  0.0108  -0.0255 -0.00189 -0.0125
        # 6 2010-06-30  0.0588  0.0310 -0.0669   0.0633  -0.0286  0.0203  -0.0644
        
        # We will use only the weights from the test set's earliest date and tack on
        # equal weights from the training set for charting purposes
        data_out <-
                training_data |>
                equalise_weights() |> 
                bind_rows(df_weights) |>
                arrange(month_end)
        
        # 10% limit check
        if(any(data_out[-1] > 0.101)) stop("One of your weights exceeds the 0.1 limit.")
        
        return(data_out)
        
}


#' Plot the total return
#' 
#' Uses the output from generate_portfolio() 
#'
#' @param df_returns a dataframe of returns
#' @param df_portfolio_weights a dataframe of portfolio weights
#' @param return_data 
#'
#' @return
#' plotly html widget
plot_total_return <- function(df_returns, df_portfolio_weights, return_data = FALSE) {
        
        returns_long <- pivot_longer(df_returns, contains("Stock"), values_to = "return")
        
        # generate equal weighted benchmark index
        benchmark_return <-
                equalise_weights(df_returns) |>
                pivot_longer(contains("Stock"), values_to = "weight") |>
                left_join(returns_long, by = c("month_end", "name")) |> 
                mutate(indexed = weight*return) |> 
                summarise(benchmark_return = sum(indexed), .by = c("month_end")) |> 
                mutate(benchmark_return = cumprod(1 + benchmark_return)*100) |> 
                pivot_longer(benchmark_return)
        
        # process the portfolio weight returns
        portfolio_return <-
                df_portfolio_weights |>
                pivot_longer(contains("Stock"), values_to = "weight") |> 
                left_join(returns_long, by = c("month_end", "name")) |> 
                mutate(indexed = weight*return) |>
                summarise(portfolio_return = sum(indexed), .by = c("month_end")) |> 
                mutate(portfolio_return = cumprod(1 + portfolio_return)*100) |> 
                pivot_longer(2)
        
        chart_data <- 
                bind_rows(benchmark_return, portfolio_return) |>
                mutate(name = stringr::str_to_title(gsub(pattern = "_", replacement = " ", x = name)))
        
        if (return_data) return(chart_data)
        
        plot <- 
                chart_data |>
                group_by(name) |>
                ggplot(aes(x = month_end, y = value, group = name, colour = name)) +
                geom_line() +
                labs(y = "Total Return", x = "Month End")
        
        ggplotly(plot)
        
}

# Run the solution --------------------------------------------------------

returns <- bind_rows(returns_train, returns_test)
portfolio_weights <- generate_portfolio(returns_train, returns_test)
plot_total_return(returns, portfolio_weights, return_data = FALSE)






plot(tsCummulative,plot.type = "s", col = c("orange", "magenta") ,lwd = c(2,2),at = "chic", format = "%Y %b", xlab = "Time", ylab = "Cummulative Returns")
# text(as.POSIXct("2009-02-28"),3.4,"Global Financial Crisis",pos = 2, cex = 1.1, srt = 90)
# text(as.POSIXct("2014-01-31"),4.9,"QE tappering",pos = 2, cex = 1.1, srt = 90)
# text(as.POSIXct("2012-08-31"),4.9,"Marikana massacre",pos = 2, cex = 1.1, srt = 90)
# text(as.POSIXct("2015-07-31"),2.5,"Anitdumping Policy",pos = 2, cex = 1.1, srt = 90)
## Add event lines
# abline(v = as.POSIXct("2009-02-28"),lwd = 3,col = "black") #GFC
# abline(v = as.POSIXct("2014-01-31"),lwd = 3,col = "black") # QE tappering
# abline(v = as.POSIXct("2012-08-31"), lwd = 3,col = "black") # Marikana massacre
# abline(v = as.POSIXct("2015-07-31"), lwd = 3,col = "black") # Antidumping
# title
title(main = "Overlapping Rolling Window Equity Curve")
# legend
# EQW, SR, BH, HRP, CM
legend("topleft",c(names(O_tsIndx)),col = c("orange","magenta"), lwd = c(2,2), lty = c('solid', 'solid'), bty = "o")